File: 2024/03/rust/src/main.rs
use regex::Regex;

fn main() {
    let input = std::fs::read_to_string("../input.txt").unwrap();
    let re = Regex::new(r"mul\((\d+),(\d+)\)|do\(\)|don't\(\)").unwrap();
    let mut result1 = 0;
    let mut result2 = 0;
    let mut enabled = true;
    for line in input.lines() {
        let captures = re.captures_iter(line);
        for it in captures {
            let v = it.get(0).unwrap().as_str();
            if v.starts_with("mul") {
                let a = it.get(1).unwrap().as_str().parse::<i64>().unwrap();
                let b = it.get(2).unwrap().as_str().parse::<i64>().unwrap();
                result1 += a * b;
                if enabled {
                    result2 += a * b;
                }
            } else if v.starts_with("do()") {
                enabled = true;
            } else if v.starts_with("don't()") {
                enabled = false;
            }
        }
    }
    println!("{result1}");
    println!("{result2}");
}

File: 2024/03/main.go
package main

import (
	"fmt"
	"os"
	"regexp"
	"strconv"
	"strings"
)

func main() {
	r, _ := regexp.Compile("mul\\((\\d+),(\\d+)\\)|do\\(\\)|don't\\(\\)")
	fileContext, _ := os.ReadFile("input.txt")
	text := string(fileContext)
	lines := strings.Split(text, "\n")
	enabled := true
	result1 := 0
	result2 := 0
	for _, line := range lines {
		if len(line) == 0 {
			continue
		}
		for _, r := range r.FindAllStringSubmatch(line, -1) {
			if strings.HasPrefix(r[0], "mul") {
				a, _ := strconv.Atoi(r[1])
				b, _ := strconv.Atoi(r[2])
				result1 += a * b
				if enabled {
					result2 += a * b
				}
			} else if strings.HasPrefix(r[0], "do()") {
				enabled = true
			} else if strings.HasPrefix(r[0], "don't()") {
				enabled = false
			}
		}
	}
	fmt.Println(result1)
	fmt.Println(result2)
}
